vllm/v1/engine/async_llm.py

```diff
@@ -536,6 +545,115 @@ async def abort(self, request_id: str | Iterable[str]) -> None:
         if self.log_requests:
             logger.info("Aborted request(s) %s.", ",".join(request_ids))
 
+    async def pause_generation(
+        self,
+        *,
+        mode: str = "force",
+        clear_cache: bool = True,
+    ) -> dict[str, Any]:
+        """Pause generation to allow model weight updates.
+
+        Args:
+            mode: ``"force"`` immediately aborts running requests (default).
+                ``"gentle"`` waits for in-flight requests to finish.
+            clear_cache: Whether to clear KV cache and prefix cache after
+                draining. Set to ``False`` to preserve cache for faster resume.
+                Default is ``True`` (clear caches).
+
+        Returns:
+            A dictionary describing the pause status, including ``cache_cleared``
+            field indicating whether caches were cleared.
+        """
+
+        if mode not in {"gentle", "force"}:
+            raise ValueError(f"Unsupported pause mode: {mode!r}")
+
+        # Use lock to ensure atomicity with resume_generation
+        async with self._pause_resume_lock:
+            async with self._pause_cond:
+                if self._is_paused:
+                    return {
+                        "paused": True,
+                        "mode": mode,
+                        "message": "Already paused",
+                        "aborted_requests": 0,
+                        "cache_cleared": False,
+                    }
+
+                self._is_paused = True
```

Comment:
Move this to the end before return?

Reply:
If we move it to the end of `pause_generation`, new generation requests will still be accepted by the engine during the execution of pause.
It's safer to set `paused` to True in the beginning of pause execution. Let me rename `_is_paused` to `_paused` to avoid status confusion. thanks